akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala

```diff
@@ -512,6 +520,20 @@ private[stream] class ActorGraphInterpreter(_initial: GraphInterpreterShell) ext
         false
     }
 
+  //this limits number of messages that can be processed synchronously during one actor receive.
+  private val shortCircuitLimit: Int = _initial.mat.settings.syncProcessingLimit
+  private var shortCircuitBuffer: FixedSizeBuffer[BoundaryEvent] = null
+  private var currentLimit: Int = shortCircuitLimit
+
+  def enqueueToShortCircuit(input: AsyncInput): Boolean = {
+    if (shortCircuitBuffer == null) shortCircuitBuffer = FixedSizeBuffer[BoundaryEvent](shortCircuitLimit)
+    if (!shortCircuitBuffer.isFull && currentLimit > 0) {
```

Comment:
While processing the inputs from the list it is random chance whether any given callback will be enqueued or not—in particular it is possible that after having sent one callback via the ActorRef the next one will go to the short circuit. This leads to a reordering or callbacks coming from within this same actor. There is no problem with external callbacks since their ordering guarantees are governed by the ActorRef semantics.